File: 1008-binary-tree-cameras/binary-tree-cameras.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    int cameras = 0;
    int f (TreeNode* node) {
        if (!node) {
            return 1;
        }
        if (!(node -> left) && !(node -> right)) {
            return 0;
        }

        int l = f(node -> left);
        int r = f(node -> right);

        if (l == 0 || r == 0) {
            cameras++;
            return 2;
        }
        if (l == 2 || r == 2) {
            return 1;
        }
        return 0;
    }
public:
    int minCameraCover(TreeNode* root) {

        // 3 states
        /*
            0 => not covered
            1 => covered
            2 => covered and has camera
        */
        if (f(root) == 0) {
            cameras++;
        }
        return cameras;


        
    }
};

File: 1008-binary-tree-cameras/binary-tree-cameras.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int cameras = 0;
    private int solve(TreeNode root) {
        if (root == null) {
            return 2;
        }

        int left = solve(root.left);
        int right = solve(root.right);

        if (left == 1 || right == 1) {
            cameras += 1;
            return 0;
        } else if (left == 0 || right == 0) {
            return 2;
        } else {
            return 1;
        }
    }
    public int minCameraCover(TreeNode root) {
        if (solve(root) == 1) {
            return cameras + 1;
        }
        return cameras;
    }
}